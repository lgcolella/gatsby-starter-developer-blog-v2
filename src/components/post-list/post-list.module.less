@import (reference) '../../style/index.less';

.container {
    .flex-center();
    flex-direction: column;
    width: 100%;
}

.post {
    .flex-center();
    flex-wrap: initial;
    flex-direction: row;
    margin-top: @margin-medium * 2;
    margin-bottom: @margin-medium * 2;
    width: 100%;
    @media @mobile {
        flex-direction: column;
    }
}

.cover {
    overflow: hidden;
    width: 40%;
    min-width: 40%;
    box-shadow: @box-shadow-cover;
    border: 1px solid @theme-text;
    @media @mobile {
        min-width: 100%;
        width: 100%;
    }
    @media @desktop {
        & > a > div {
            transition: transform @transition-time;
            &:hover {
                transform: scale(1.1);
            }
        }
    }
}

.content {
    padding-left: @padding-medium * 4;
    h3 {
        margin: @margin-zero;
        text-align: left;
        text-shadow: @text-shadow-title;
    }
    p {
        margin: @margin-small @margin-zero;
        padding: @padding-small @padding-zero;
    }
    @media @mobile {
        padding-top: @padding-medium;
        padding-left: @padding-zero;
    }
    @media @desktop {
        h3 {
            .underline-animation-complete(@theme-title, 3px, @margin-small);
        }
    }
}
