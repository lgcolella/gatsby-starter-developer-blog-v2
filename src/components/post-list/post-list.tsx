/* Vendor imports */
import React from 'react'
import PropTypes from 'prop-types'
import { Link } from 'gatsby'
import Img from 'gatsby-image'
/* App imports */
import { ChildImageSharpFluid, AvailableTag } from '../../types'
import TagList from '../tag-list'
import Utils from '../../utils'
const style = require('./post-list.module.less')

const PostList = ({ posts, contentClassName }: Props) => (
  <div className={style.container}>
    {posts.map((post, index) => {
      const { title, date, path, tags, cover, excerpt } = post.node.frontmatter
      return (
        <div key={title} className={style.post}>
          <div className={style.cover}>
            <Link to={Utils.resolveUrl(path)}>
              <Img
                fluid={cover.childImageSharp.fluid}
                title={title}
                alt={title}
              />
            </Link>
          </div>
          <div className={[style.content, contentClassName].join(' ')}>
            <Link to={Utils.resolveUrl(path)}>
              {date ? <label>{date}</label> : null}
              <h3>{title}</h3>
              <p>{excerpt}</p>
            </Link>
            <TagList tags={tags} />
          </div>
        </div>
      )
    })}
  </div>
)

PostList.propTypes = {
  posts: PropTypes.arrayOf(
    PropTypes.shape({
      node: PropTypes.shape({
        frontmatter: PropTypes.shape({
          title: PropTypes.string.isRequired,
          date: PropTypes.string,
          path: PropTypes.string.isRequired,
          tags: PropTypes.arrayOf(PropTypes.string).isRequired,
          excerpt: PropTypes.string.isRequired,
          cover: PropTypes.shape({
            childImageSharp: PropTypes.shape({
              fluid: PropTypes.object.isRequired,
            }).isRequired,
          }).isRequired,
        }),
      }),
    })
  ),
  contentClassName: PropTypes.string,
}

PostList.defaultProps = {
  contentClassName: '',
}

export interface Props {
  posts: Array<{
    node: {
      frontmatter: {
        title: string
        date?: string
        path: string
        tags: AvailableTag[]
        excerpt: string
        cover: {
          childImageSharp: ChildImageSharpFluid
        }
      }
    }
  }>
  contentClassName?: string
}

export default PostList
