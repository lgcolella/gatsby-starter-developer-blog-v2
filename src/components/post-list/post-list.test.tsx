/* Vendor imports */
import React from 'react'
import { shallow } from 'enzyme'
/* App imports */
import Component from './index'
import { Props } from './post-list'
import { ImageSharpFluid } from '../../../tests/mocks'

describe('Post List component', () => {
  const mockPostList: Props['posts'] = [
    // Article #1
    {
      node: {
        frontmatter: {
          title: 'article #1',
          date: '2019-02-13',
          path: '/article-1-path',
          tags: ['javascript', 'nodejs', 'angular', 'react'],
          excerpt: 'excerpt of the article 1',
          cover: {
            childImageSharp: ImageSharpFluid,
          },
        },
      },
    },
    // Article #2
    {
      node: {
        frontmatter: {
          title: 'article #2',
          path: '/article-2-path',
          tags: ['javascript', 'laravel', 'nodejs', 'angular'],
          excerpt: 'excerpt of the article 2',
          cover: {
            childImageSharp: ImageSharpFluid,
          },
        },
      },
    },
  ]

  test('matches snapshot', () => {
    let wrapper = shallow(<Component posts={mockPostList} />)

    expect(wrapper).toMatchSnapshot()
  })
})
