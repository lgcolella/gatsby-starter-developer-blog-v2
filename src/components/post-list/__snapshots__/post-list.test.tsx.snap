// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post List component matches snapshot 1`] = `
<div
  className="container"
>
  <div
    className="post"
    key="article #1"
  >
    <div
      className="cover"
    >
      <mockConstructor
        to="/article-1-path/"
      >
        <Image
          Tag="div"
          alt="article #1"
          durationFadeIn={500}
          fadeIn={true}
          fluid={
            Object {
              "aspectRatio": 1.7,
              "sizes": "(max-width: 600px) 100vw, 600px",
              "src": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
              "srcSet": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
            }
          }
          loading="lazy"
          title="article #1"
        />
      </mockConstructor>
    </div>
    <div
      className="content "
    >
      <mockConstructor
        to="/article-1-path/"
      >
        <label>
          2019-02-13
        </label>
        <h3>
          article #1
        </h3>
        <p>
          excerpt of the article 1
        </p>
      </mockConstructor>
      <TagList
        tags={
          Array [
            "javascript",
            "nodejs",
            "angular",
            "react",
          ]
        }
      />
    </div>
  </div>
  <div
    className="post"
    key="article #2"
  >
    <div
      className="cover"
    >
      <mockConstructor
        to="/article-2-path/"
      >
        <Image
          Tag="div"
          alt="article #2"
          durationFadeIn={500}
          fadeIn={true}
          fluid={
            Object {
              "aspectRatio": 1.7,
              "sizes": "(max-width: 600px) 100vw, 600px",
              "src": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
              "srcSet": "/static/da8d0fc0af7f0d310517048e303b3fcf/a5a41/preview.jpg",
            }
          }
          loading="lazy"
          title="article #2"
        />
      </mockConstructor>
    </div>
    <div
      className="content "
    >
      <mockConstructor
        to="/article-2-path/"
      >
        <h3>
          article #2
        </h3>
        <p>
          excerpt of the article 2
        </p>
      </mockConstructor>
      <TagList
        tags={
          Array [
            "javascript",
            "laravel",
            "nodejs",
            "angular",
          ]
        }
      />
    </div>
  </div>
</div>
`;
